### ISSUE TITLE
Loss explodes problem

### ISSUE THREAD

**Issue by talkingwallace** [zzd1992/GBDTMO#5]
Hi~, I try to train a GBDTMO model on a large multi-class dataset 'covtype'. The 'covtype' is from libsvm, this dataset contains 7 labels and 580000+ samples.
As the figures shown below, the loss value decreases normally at the very beginning epochs but starts to rise at epoch 21 and finally reaches a very large value at 99 epoch. 
![企业微信截图_16280584876610](https://user-images.githubusercontent.com/10768970/128132515-8824af68-da88-4c54-80d1-d3aca71767b2.png)
![企业微信截图_16280584975230](https://user-images.githubusercontent.com/10768970/128132523-75e81857-4de3-4046-8641-6a793d2f1ddd.png)

This is the final accuracy.
![企业微信截图_16280585232079](https://user-images.githubusercontent.com/10768970/128132551-756e7cbd-d503-420f-95be-7a9fe523e02d.png)

I checked the output model file and found out that some trees have very large output weights. I think perhaps this is the cause of the loss explosion.

I packed the data and the running script together and you can download it through this share link:
https://www.jianguoyun.com/p/DY1tnQ0QruzbCRj89YUE

Could you please check this problem? I think we can discuss together. :D

**Comment by zzd1992** [comment]
Loss explosion may happen when we use cross-entropy for training.
Using larger `min_samples` and `reg_l2` is helpful to avoid this.
For example, you can try `min_samples=8` and `reg_l2=1.0`.
A more straightforward solution is to clip the weights of trees. This is used in XGBoost.
But weights clipping is not implemented in this project currently.

**Comment by talkingwallace** [comment]
Ok, I will try, thanks for your reply
